specifyweb/frontend/js_src/lib/components/WbPlanView/mappingHelpers.ts

```diff
@@ -75,14 +94,44 @@ export const formattedEntry = `${schema.fieldPartSeparator}formatted`;
  */
 export const emptyMapping = '0';
 
+/**
+ * Returns a complete tree definition name from a tree definition name
+ * (e.x Taxononomy => %Taxonomy)
+ * Opposite of getNameFromTreeDefinitionName
+ */
+export const formatTreeDefinition = (definitionName: string): string =>
+  `${schema.treeDefinitionSymbol}${definitionName}`;
+
 /**
  * Returns a complete tree rank name from a tree rank name
  * (e.x Kingdom => $Kingdom)
  * Opposite of getNameFromTreeRankName
  *
  */
 export const formatTreeRank = (rankName: string): string =>
-  `${schema.treeSymbol}${rankName}`;
+  `${schema.treeRankSymbol}${rankName}`;
+
+// Delimiter used for rank name keys i.e: <rankname>~><treeId>
+const RANK_KEY_DELIMITER = '~>';
+
+/**
```

Comment:
if we have this, do we even need the treeDefinitionSymbol?

can we simplify things by adding a tree def id to every usage of a tree rank?

Reply:
once we add these symbols, it would be very hard to remove them as it would break previously creating upload plans.

if i am not wrong, it appears that this PR introduces two new mapping path symbols, when one would suffice.

Reply:
We already do pass a treedefId to every usage of a tree rank. There was an issue where if the mapping has different trees with the same rank name, it would generate an incorrect upload plan because the key name would be the same for all mappings with that rank name and so the key would just get overwritten.

More info on that in these comments: 
https://github.com/specify/specify7/pull/5091#issuecomment-2327526079
https://github.com/specify/specify7/pull/5091#issuecomment-2328037741

`RANK_KEY_DELIMITER` is not really a part of the mapping but rather a part of the building/parsing the upload plan. Maybe placing it in `mappingHelpers.ts` was a bad idea and should get moved to `uploadPlanBuilder.ts`?

Other than that, I'm not sure we could use the same symbols for mapping and for generating unique key names as those are different usages in my opinion

Reply:
For the usage of this, the test file for the uploadplan provides an example. 

https://github.com/specify/specify7/blob/10324c97095b34c296b37d3a18a354dba7a95e6c/specifyweb/frontend/js_src/lib/tests/fixtures/uploadplan.1.json#L276-L322

Here are my thoughts on this: 

This approach _does_ make it pretty simple to parse the tree definition id from the ranks, but I think it could be made a little more intuitive.

https://github.com/specify/specify7/blob/10324c97095b34c296b37d3a18a354dba7a95e6c/specifyweb/workbench/upload/treerecord.py#L137-L146

I think if we are modifying the upload plan this much already, it may be worthwhile to format the upload plan in a way which more closely resembles the added functionality. 

Particularly, I think an "inheritance" approach to mapping would make the most sense. 
Something like the following would be the most ideal I think: 

```json
{
   "ranks":{
      "Taxonomy":{
         "Class":{
            ...
         },
         "treeId":1 # You can keep treeId here if you want, but we should be able to identify the treeDef by name as well
      },
      "2" :{ # Or, maybe a little less intuitive: just use the treeDefId as a key 🤷 
         "Class":{
            ...
         }, 
         "treeId": 2 # and remove this
      }
   }
}
```

Of course, the implementation of this would be far more difficult to pull off so it doesn't have to follow the above exactly (`ranks` could be an array for example), but "it would be very hard to remove them as it would break previously creating upload plans" is a real concern (https://github.com/specify/specify7/pull/5091#discussion_r1751222126).

We are kind of repeating ourselves here as well, which can make things even more confusing: the treeDefId is in two places in the upload plan and _only_ needed in the rank name to resolve duplicate rank names across trees: which seems more like a design problem at odds with the newer approach than anything else.

Reply:
Actually, it's much more simple to make it an array of trees.

Reply:
Shouldn't be too difficult to make it an array on the frontend